### ISSUE TITLE
no error on invalid argument to `dataclass_transform`

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#4170]
### Summary

```py
from typing import dataclass_transform


@dataclass_transform(asdfsadf=True) # no error
def dataclass[T: type](t: T) -> T:
    return t

```

[Playground](https://play.ty.dev/a8dba796-58e4-48b9-8633-a0d4f8ee524e)

from [the spec](https://typing.python.org/en/latest/spec/dataclasses.html#dataclass-transform-parameters):

> - `kwargs` allows arbitrary additional keyword args to be passed to `dataclass_transform`. This gives type checkers the freedom to support experimental parameters without needing to wait for changes in `typing.py`. Type checkers should report errors for any unrecognized parameters.

(related: #2666)

### Version

7fd30bb81

**Comment by sharkdp** [comment]
Thank you for reporting this.

I see the idea behind it, but I don't like that typeshed adds `**kwargs: Any` here. Maybe we should just patch this in our vendored stubs (i.e. just remove that parameter). @AlexWaygood what do you think?

I personally think it should be the other way around: if type checkers want to experiment with extra arguments to `dataclass_transform`, they should modify `dataclass_transform`. Everyone who doesn't experiment should get the correct signature for that function.

**Comment by AlexWaygood** [comment]
The runtime signature has a `**kwargs` parameter here, so strictly speaking typeshed is just being accurate to the runtime signature here:

```pycon
>>> from typing import dataclass_transform
>>> help(dataclass_transform)
Help on function dataclass_transform in module typing:

dataclass_transform(
    *,
    eq_default: bool = True,
    order_default: bool = False,
    kw_only_default: bool = False,
    frozen_default: bool = False,
    field_specifiers: tuple[type[Any] | Callable[..., Any], ...] = (),
    **kwargs: Any
) -> _IdentityCallable
    Decorator to mark an object as providing dataclass-like behaviour.

    The decorator can be applied to a function, class, or metaclass.

    Example usage with a decorator function::

        @dataclass_transform()
        def create_model[T](cls: type[T]) -> type[T]:
            ...
            return cls

        @create_model
        class CustomerModel:
            id: int
            name: str

    On a base class::

        @dataclass_transform()
        class ModelBase: ...

        class CustomerModel(ModelBase):
            id: int
            name: str

    On a metaclass::

        @dataclass_transform()
        class ModelMeta(type): ...

        class ModelBase(metaclass=ModelMeta): ...
```

And typeshed's CI checks that verify consistency with the runtime signature would fail if typeshed got rid of the `**kwargs` parameter in the stub.

That said, it's pretty easy to add exemptions to that CI check in situations where typeshed deliberately tells "white lies" that don't match the runtime; typeshed does that in lots of other places. So it's very possible that we could have this changed upstream in typeshed. I think your rationale makes as much sense for other type checkers as it does for us.

Since `dataclass_transform` is obviously quite heavily special-cased by type checkers, though, it's a change where we'd probably want to check that it had no adverse affects on inference for (at least) mypy, pyright and pyrefly before merging it upstream. So in this case, I have no objections to patching typeshed in our vendored copy in the meantime.
